dll/win32/wdmaud.drv/legacy.c

```diff
@@ -1,89 +1,306 @@
 /*
- * PROJECT:     ReactOS Sound System
- * LICENSE:     GPL - See COPYING in the top level directory
- * FILE:        dll/win32/wdmaud.drv/wdmaud.c
- *
- * PURPOSE:     WDM Audio Driver (User-mode part)
- * PROGRAMMERS: Andrew Greenwood (silverblade@reactos.org)
-                Johannes Anderwald
- *
- * NOTES:       Looking for wodMessage & co? You won't find them here. Try
- *              the MME Buddy library, which is where these routines are
- *              actually implemented.
- *
+ * PROJECT:     ReactOS Audio Subsystem
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     WDM Audio Driver Mapper (User-mode part)
+ * COPYRIGHT:   Copyright 2007-2008 Andrew Greenwood (silverblade@reactos.org)
+ *              Copyright 2009-2010 Johannes Anderwald
+ *              Copyright 2022 Oleg Dubinskiy (oleg.dubinskij30@gmail.com)
  */
 
 #include "wdmaud.h"
 
-#define NDEBUG
-#include <debug.h>
-#include <mmebuddy_debug.h>
+#include <mmixer.h>
 
-#define KERNEL_DEVICE_NAME      L"\\\\.\\wdmaud"
+#define YDEBUG
+#include <debug.h>
 
+extern BOOL CriticalSectionInitialized;
+extern CRITICAL_SECTION WdmAudCriticalSection;
+extern MIXER_CONTEXT MixerContext;
 HANDLE KernelHandle = INVALID_HANDLE_VALUE;
 DWORD OpenCount = 0;
 
+/*
+    This is a wrapper around DeviceIoControl which provides control over
+    instantiated sound devices. It waits for I/O to complete (since an
+    instantiated sound device is opened _In_ overlapped mode, this is necessary).
+*/
+MMRESULT
+WdmAudIoControl(
+    _In_ PWDMAUD_DEVICE_INFO DeviceInfo,
+    _In_opt_ ULONG BufferSize,
+    _In_opt_ PVOID Buffer,
+    _In_ DWORD IoControlCode)
+{
+    MMRESULT Result = MMSYSERR_NOERROR;
+    OVERLAPPED Overlapped;
+    DWORD Transferred = 0;
+    BOOL IoResult;
+
+    /* Overlapped I/O is done here - this is used for waiting for completion */
+    ZeroMemory(&Overlapped, sizeof(OVERLAPPED));
+    Overlapped.hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    if ( ! Overlapped.hEvent )
+        return Win32ErrorToMmResult(GetLastError());
+
+    if (DeviceInfo->DeviceType != AUX_DEVICE_TYPE &&
+        DeviceInfo->DeviceType != MIXER_DEVICE_TYPE)
+    {
+        /* Enter critical section */
+        EnterCriticalSection(&WdmAudCriticalSection);
```

Comment:
What exactly does this critical section protect? You seem to acquire it for just about everything.

Reply:
I did this previously because Windows XP/2003 also manages a critical section in wdmaud.drv. But if it's not necessarily, I removed it now. :slightly_smiling_face: